python/src/trezorlib/tools.py

```diff
@@ -195,42 +222,74 @@ def normalize_nfc(txt):
     return unicodedata.normalize("NFC", txt).encode()
 
 
-class expect:
-    # Decorator checks if the method
-    # returned one of expected protobuf messages
-    # or raises an exception
-    def __init__(self, expected, field=None):
-        self.expected = expected
-        self.field = field
+# NOTE for type tests (mypy/pyright):
+# Overloads below have a goal of enforcing the return value
+# that should be returned from the original function being decorated
+# while still preserving the function signature (the inputted arguments
+# are going to be type-checked).
+# Currently (November 2021) mypy does not support "ParamSpec" typing
+# construct, so it will not understand it and will complain above
+# definitions below.
 
-    def __call__(self, f):
+
+@overload
+def expect(
+    expected: "Type[MT]",
+) -> "Callable[[Callable[P, MessageType]], Callable[P, MT]]":
+    ...
+
+
+@overload
+def expect(
+    expected: "Type[MT]", *, field: str, ret_type: "Type[R]"
+) -> "Callable[[Callable[P, MessageType]], Callable[P, R]]":
+    ...
+
+
+def expect(
+    expected: "Type[MT]",
+    *,
+    field: Optional[str] = None,
+    ret_type: "Optional[Type[R]]" = None,
+) -> "Callable[[Callable[P, MessageType]], Union[Callable[P, MT], Callable[P, R]]]":
+    """
+    Decorator checks if the method
+    returned one of expected protobuf messages
+    or raises an exception
+    """
+
+    def decorator(f: "Callable[P, MessageType]") -> "Callable[P, Union[MT, R]]":
         @functools.wraps(f)
-        def wrapped_f(*args, **kwargs):
+        def wrapped_f(*args: "P.args", **kwargs: "P.kwargs") -> "Union[MT, R]":
             __tracebackhide__ = True  # for pytest # pylint: disable=W0612
             ret = f(*args, **kwargs)
-            if not isinstance(ret, self.expected):
-                raise RuntimeError(f"Got {ret.__class__}, expected {self.expected}")
-            if self.field is not None:
-                return getattr(ret, self.field)
+            if not isinstance(ret, expected):
+                raise RuntimeError(f"Got {ret.__class__}, expected {expected}")
+            if field is not None:
+                return getattr(ret, field)
             else:
                 return ret
 
         return wrapped_f
 
+    return decorator
+
 
-def session(f):
+def session(f: "FF") -> "FF":
     # Decorator wraps a BaseClient method
     # with session activation / deactivation
+    # With "FF" we are enforcing that for type-checking purposes the return value
+    # of this decorator will be the same as the decorated function
     @functools.wraps(f)
-    def wrapped_f(client, *args, **kwargs):
+    def wrapped_f(client: "TrezorClient", *args: Any, **kwargs: Any) -> "MessageType":
```

Comment:
the `-> MessageType` here is probably wrong, should be `Any`, that might also solve the type:ignore below

Reply:
When changed to return `Any`, the `pyright` error type-ignored below remains, and is the same (just `MessageType` changed to `Any`): 
```
Expression of type "(client: TrezorClient, *args: Any, **kwargs: Any) -> Any" cannot be assigned to return type "FF@session"
    Type "(client: TrezorClient, *args: Any, **kwargs: Any) -> Any" cannot be assigned to type "FF@session" 
```

However, it is right that not all `@session`-decorated functions return `MessageType`, so I will change it to `Any`

Reply:
Changed in 4cfe54a

Reply:
I see!
the issue here is that we don't accept functions of _any arguments_, but functions whose first argument is `client`. ParamSpec might be able to express that?

Reply:
Playing with `ParamSpec` is interesting, but it still has some problems, for example:
```python
P = ParamSpec("P")
def session(f: "Callable[P, Any]") -> "Callable[P, Any]":
    def wrapped_f(client: "TrezorClient", *args: "P.args", **kwargs: "P.kwargs") -> Any:
```
complains about `Expected 0 positional arguments` at the place where we call `return f(client, *args, **kwargs)`

Reply:
When using `Concatenate`, the same way as used in `with_client` decorator (`Callable[Concatenate[TrezorClient, P], R]`), it was fine in this decorator itself, but it uncovered some other interesting things. 

The usages of the `session`-decorated functions are complaining about `Expected 0 (1, 3...) positional arguments` and maybe more importantly `Argument of type "TrezorClient" cannot be assigned to parameter "label" of type "str | None"`.

There seems to be a mismatch that we are supplying the `TrezorClient` into the function when we call it, but it is also being added by the `@session` decorator.

Reply:
Finally resolved in ec223c0. The trick was in supplying the same `Callable[Concatenate[TrezorClient, P], Any]` on the argument type and return type